benchflow-test-manager/application/src/main/java/cloud/benchflow/testmanager/scheduler/running/RunningStatesHandler.java

```diff
@@ -0,0 +1,370 @@
+import cloud.benchflow.testmanager.tasks.running.DetermineExecuteInitialValidationSetTask;
+import cloud.benchflow.testmanager.tasks.running.DetermineExplorationStrategyTask;
+import cloud.benchflow.testmanager.tasks.running.HandleExperimentResultTask;
+import cloud.benchflow.testmanager.tasks.running.RemoveNonReachableExperimentsTask;
+import cloud.benchflow.testmanager.tasks.running.ValidatePredictionFunctionTask;
+import cloud.benchflow.testmanager.tasks.running.ValidateTerminationCriteria;
+import cloud.benchflow.testmanager.tasks.running.ValidateTerminationCriteria.TerminationCriteriaResult;
+import java.util.concurrent.ConcurrentMap;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Future;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * @author Jesper Findahl (jesper.findahl@gmail.com) created on 2017-07-02
+ */
+public class RunningStatesHandler {
+
+  /**
+   * Every running state checks if the test has been terminated after each task
+   * to avoid that the test state is overwritten.
+   */
+
+  private static Logger logger =
+      LoggerFactory.getLogger(RunningStatesHandler.class.getSimpleName());
+
+  private ConcurrentMap<String, Future> testTasks;
+
+  private ExecutorService taskExecutorService;
+
+  private ExplorationModelDAO explorationModelDAO;
+  private BenchFlowTestModelDAO testModelDAO;
+
+  private TestTaskScheduler testTaskScheduler;
+
+  public RunningStatesHandler(ConcurrentMap<String, Future> testTasks,
+      ExecutorService taskExecutorService, TestTaskScheduler testTaskScheduler) {
+
+    this.testTasks = testTasks;
+    this.taskExecutorService = taskExecutorService;
+    this.testTaskScheduler = testTaskScheduler;
+  }
+
+  public void initialize() {
+
+    this.explorationModelDAO = BenchFlowTestManagerApplication.getExplorationModelDAO();
+    this.testModelDAO = BenchFlowTestManagerApplication.getTestModelDAO();
+
+  }
+
+  public void determineExplorationStrategy(String testID) {
+
+    logger.info("determineExplorationStrategy with testID: " + testID);
+
+    DetermineExplorationStrategyTask task = new DetermineExplorationStrategyTask(testID);
+
+    // TODO - should go into a stateless queue (so that we can recover)
+    Future<?> future = taskExecutorService.submit(task);
+
+    // replace with new task
+    testTasks.put(testID, future);
+
+    waitForRunningTaskToComplete(testID, future, ADD_STORED_KNOWLEDGE);
+  }
+
+  public void addStoredKnowledge(String testID) {
+
+    logger.info("addStoredKnowledge with testID: " + testID);
+
+    AddStoredKnowledgeTask task = new AddStoredKnowledgeTask(testID);
+
+    // TODO - should go into a stateless queue (so that we can recover)
+    Future<?> future = taskExecutorService.submit(task);
+
+    // replace with new task
+    testTasks.put(testID, future);
+
+    try {
+
+      // wait for task to complete
+      future.get();
```

Comment:
If you make task cancellable,  then probably here you are going to get an [exception](https://docs.oracle.com/javase/7/docs/api/java/util/concurrent/CancellationException.html) and you follow the exceptional flow, so that you do not need to check all the time if something is terminated. When you apply the previous comment, then the handling can be delegated to a method.